retryable_http_client.go

```diff
@@ -0,0 +1,63 @@
+package statuspage
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+func NewRetryableClient() *retryablehttp.Client {
+	client := retryablehttp.NewClient()
+	client.CheckRetry = retryPolicy
+	client.Backoff = backoffPolicy
+	return client
+}
+
+const (
+	StatusRateLimitExceeded = 420
+)
+
+// We're wrapping original retryablehttp.DefaultRetryPolicy and adding retry on 420 HTTP code
+func retryPolicy(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	defaultShouldRetry, err := retryablehttp.DefaultRetryPolicy(ctx, resp, err)
+	if err != nil {
+		return false, err
+	}
+
+	isRateLimited := resp != nil && resp.StatusCode == StatusRateLimitExceeded
+	shouldRetry := defaultShouldRetry || isRateLimited
+	return shouldRetry, nil
+}
+
+// We're wrapping original retryablehttp.DefaultBackoff and using response header in such format `Retry-After: 60`
+// retryablehttp already implements this behaviour, but it's executed only for 429 and 503 HTTP codes, and we need it for 420 HTTP code as well
+func backoffPolicy(min, max time.Duration, attemptNum int, resp *http.Response) time.Duration {
+	if resp != nil {
+		if resp.StatusCode == StatusRateLimitExceeded {
+			if sleep, err := parseRetryAfterHeader(resp.Header.Get("Retry-After")); err != nil {
```

Comment:
Do you know whether Statuspage returns this header? The behaviour is undocumented.

Reply:
Here's real 420 response using `curl` during rate limiting

```
> GET /v1/pages/XXX/components/XXX HTTP/2
> Host: api.statuspage.io
> User-Agent: curl/8.1.2
> Accept: */*
> Authorization: OAuth XXX
>
< HTTP/2 420
< date: Tue, 01 Oct 2024 13:06:17 GMT
< content-type: application/json
< retry-after: 60
< vary: Origin, Accept-Encoding
< cache-control: no-cache
< x-runtime: 0.016356
< server: AtlassianEdge
< x-content-type-options: nosniff
< x-xss-protection: 1; mode=block
< atl-traceid: XXX
< strict-transport-security: max-age=63072000; preload
< report-to: {"endpoints": [{"url": "https://dz8aopenkvv6s.cloudfront.net"}], "group": "endpoint-1", "include_subdomains": true, "max_age": 600}
< nel: {"failure_fraction": 0.001, "include_subdomains": true, "max_age": 600, "report_to": "endpoint-1"}
< server-timing: atl-edge;dur=89,atl-edge-internal;dur=3,atl-edge-upstream;dur=87,atl-edge-pop;desc="aws-us-east-1"
<
* Connection #0 to host api.statuspage.io left intact
{"error":"Too many requests, enhance your calm"}%
```

As you can see response code is 420 and `retry-after: 60` header is in place. 
Here's where they mention `retry-after`, but unfortunately it's not official docs: https://community.atlassian.com/t5/Statuspage-articles/Upcoming-rate-limiting-changes-to-Statuspage-REST-API/ba-p/1455362

Btw `retry-after` is an rfc7231 header, but it's not documented on the Statuspage API side. They only provide such note:
>Each API token is limited to 1 request / second as measured on a 60 second rolling window. To get this limit increased, please contact us at https://support.atlassian.com/contact

> Error codes 420 or 429 indicate that you have exceeded the rate limit and the request has been rejected.

in https://developer.statuspage.io/#section/Rate-Limiting

so maybe if there's not `retry-after` header or we cannot parse it, then in case of 420 HTTP response we should wait 60s by default?